sale_order_product_assortment/models/sale_order.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2020 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    whitelist_product_ids = fields.Many2many(
+        comodel_name='product.product',
+        string='Whitelisted Products',
+        compute='_compute_product_assortment_ids',
+    )
+    blacklist_product_ids = fields.Many2many(
+        comodel_name='product.product',
+        string='Blacklisted Products',
+        compute='_compute_product_assortment_ids',
+    )
+    has_whitelist = fields.Boolean(
+        default='False', compute='_compute_product_assortment_ids')
+    has_blacklist = fields.Boolean(
+        default='False', compute='_compute_product_assortment_ids')
+
+    @api.depends('partner_id')
+    def _compute_product_assortment_ids(self):
+        if self.partner_id:
+            filters = self.env['ir.filters'].search([
+                (
+                    'partner_ids',
+                    'in',
+                    (self.partner_id + self.partner_id.parent_id).ids
+                ),
+            ])
+            whitelist_products = set()
+            blacklist_products = set()
+            for fil in filters:
+                whitelist_products = whitelist_products.union(
+                    fil.whitelist_product_ids.ids)
```

Comment:
Recordsets allow to operations too.
See https://www.odoo.com/documentation/12.0/reference/orm.html#set-operations

Reply:
This is done on purpose for optimization.